### ISSUE TITLE
[css-pseudo] Allow custom Properties in ::selection pseudos (revisited)

### ISSUE THREAD

**Issue by schenney-chromium** [w3c/csswg-drafts#11317]
Chromium continues to try to ship the Highlight Inheritance model for selection. The third attempt has failed because CSS pre-processing and generation tools create `::selection` rules like this:
```html
div::selection {
   --my-opacity: 1;
   background-color: rgb(0 0 0/var(--my-opacity));
}
```
In https://github.com/w3c/csswg-drafts/issues/6264 we allowed custom properties on highlight pseudos.

Then in  https://github.com/w3c/csswg-drafts/issues/6641 we also said that all custom properties should be copied from the root for the `root::selection`, and then start adding those in the `::selection` inheritance path.

This proved insufficient when launched in Chromium because many sites rely on `::selection` inheriting custom properties from their originating elements. So we discussed disallowing custom properties in highlights and inheriting from the originating elements, and resolved to do that in https://github.com/w3c/csswg-drafts/issues/9909.

@tabatkins commented in https://github.com/w3c/csswg-drafts/issues/6641 that the example above must be allowed for their support. @LeaVerou also raised the same concern in one of the issues above. I did not pay enough attention to that. So the change from #9909 broke many many major sites.

To make it possible to ship highlight inheritance for `::selection`, and bring `::selection` implementations into line with other highlight pseudos, I believe we need this ...

*`::selection` pseudo elements inherit custom properties from their `originating-element` and then apply custom properties from the `::selection` rule itself. Custom properties from `::selection` rules apply only within that rule, and are not inherited by descendants.*

My current thinking is that we limit this behavior to `::selection` and not other highlight pseudos, as a back-compatibility requirement. It's not a strongly held position.

Proposing async resolution but feel free to add to the agenda. 4 week time limit given holidays.

@delan @fantasai @frivoal

**Comment by schenney-chromium** [comment]
I should note that implementing the proposed behavior is likely straight-forward and efficient (it's a 2 line change in Chromium). Implementing dual inheritance paths, through both originating and highlight trees, is likely very bad for performance,

**Comment by astearns** [comment]
I am a bit concerned about making `::selection` work differently than other highlight pseudos. As use of the other pseudos rises, won’t people be creating similar rules for them with pre-processors and expect those to work like `::selection`?

**Comment by schenney-chromium** [comment]
I'm totally fine with making it apply to all highlight pseudos.

So the proposed resolution would be "Allow custom properties in highlight pseudos but they (a) are not inherited and (b) override custom properties with matching names taken from the originating element."

**Comment by LeaVerou** [comment]
> I am a bit concerned about making `::selection` work differently than other highlight pseudos. As use of the other pseudos rises, won’t people be creating similar rules for them with pre-processors and expect those to work like `::selection`?

Indeed. I believe all issues with `::selection` also apply to highlight pseudos and should be fixed there too.

**Comment by schenney-chromium** [comment]
> So the proposed resolution would be:
> 
> `Highlight pseudo elements inherit custom properties from their originating-element and then apply custom properties from the highlight itself. Custom properties in Highlight Pseudos are not inherited through the highlight inheritance chain.`

Final call for objections to this proposed resolution. I'll get a spec PR up within the next week where folks will also be able to comment.

**Comment by andruud** [comment]
Having a brief look at the PR, what is the model we're going for here? Custom properties apply to the pseudo (thus affecting what the computed value is), yet they "do not inherit"? Does this mean that a pseudo-element element style does not inherit _computed values_ from its parent style, but instead some other kind of value? This would be very annoying indeed, since nothing (else) would need to hold on to such alternate-reality-computed-values, as far as I can tell. We'd need to store/manage those values alongside the actual computed values, just for this.

Why can we not just let them inherit, again?

---

Sidenote: I do think it's intended that the label `Async Resolution: Call For Consensus` is applied by CSSWG chairs only.

I also don't think this is a great candidate for an async resolution---at least not in its current state.

**Comment by schenney-chromium** [comment]
The correct proposed language is "All custom properties inherit from the originating element" combined with re-allowing custom properties on the highlight itself. Much of the confusion here results from my relative inexperience writing spec PRs.

The proposal is the former resolution, Issue #9909, plus re-allowing custom properties on the highlight itself. The former resolution both banned custom properties on the highlight itself _and_ said inherit from the originating element.

The current spec draft is in fact wrong in that it says both "custom properties are inherited from the originating element" and that they inherit through the highlight cascade. I just updated the PR to fix this. https://github.com/w3c/csswg-drafts/pull/11528

In practice this means custom properties use the long-standing back-compatible browser behavior, while other properties move to the new model. This is done because so much existing selection styling depends on custom properties, and without this custom-property behavior we can never launch the improved highlight inheritance model. We've tried to launch 3 times in Chromium and this issue has tripped us up every time. The current proposal addresses all of the known breaking issues around custom properties (by basically reverting to the old behavior).

For those following along:
* Launch attempt 1 failed because custom properties defined on the root element, as was done on GitHub, for example, were not applying to the selection at all. So we said that the root highlight inherits custom properties from the root element.
* Launch attempt 2 failed because there were developer focused sites telling authors to use custom property inheritance in the element cascade to achieve sensible selection property inheritance. That was great advice, but only copying properties from the root then inheriting through highlights broke that. So we said "no custom properties on highlights at all, just take from originating". Various people were unhappy about the "no custom properties on highlights at all" portion of this but I stupidly did not follow through in investigating the prevalence.
* Launch attempt 3 failed because a widely used CSS package did in fact require custom properties defined on highlights.

Hence this issue and proposed language, which addresses all three failure modes.

From an implementation perspective, the highlight gets a shared copy of all the originating element's variables and then adds in any that the highlight defines. So basically the same as regular custom property inheritance.

If for consistency we really want to have custom properties inherit through the highlight cascade we would need to split selection out specifically, which is argued against above, or support both methods until sites can migrate (i.e. for years). Supporting both requires two sets of custom properties in pseudo highlights and some method for resolving conflicts. That really is something I do not want to do.

**Comment by schenney-chromium** [comment]
If we do want to support the back-compat behavior and the new highlight inheritance model at the same time, I think we would need a zig-zag or zipper inheritance model. Say we have
```
 root { .. }
   root::selection { ... }
parent { ... }
  parent::selection { ... }
child { ... }
  child::selection { ... }
```
then `child::selection`, in order, looks for variables on
```
child::selection (itself)
child (the originating element)
parent::selection (it's highlight parent)
parent (it's originating element's parent)
root::selection (one more level)
root (final option)
```
This sort-of supports both authors trying to use the new model and authors trying to use the old model for selection property inheritance. In particular, it will be back-compatible with custom properties inheriting from the originating element, because that's the first place to look for a property (second if you count the pseudo itself). And anything that works with previous browsers must have any properties used for selection inheriting up through the element tree.

It is possible to support both old and new behavior at the same time in one page if you (a) use property names specific to the highlights e.g. `-selection-background-color` and (b) mirror all highlight related properties between `foo` and `foo::selection` (for example). This is probably a big win, if authors get it right.

I believe then we can choose between two options, each of which would enable Highlight Inheritance to launch:
 1. Define custom properties to inherit from the originating element (and implicitly on through the element tree). The up side is that it is very easy to state and likely reasonable to implement in all browsers. The downside is that it makes custom properties inherit differently to every other property on highlight pseudos, which authors in the future will no doubt find very confusing.
 2. Define custom properties to inherit alternating between highlight pseudos and originating elements up the tree. The up side is that old content will still work while new content can use a consistent model for highlight inheritance of all properties. The down side is that it is tricky to explain and if you break the rules the behavior may be odd. And maybe it's not very efficient to implement in all browsers.

I am leaning toward option 2 because it supports back-compat while giving a viable transition to the new highlight inheritance model.

**Comment by andruud** [comment]
> From an implementation perspective, the highlight gets a shared copy of all the originating element's variables and then adds in any that the highlight defines.

Right, OK. Stated that way, it's not too bad. Got confused earlier.

> The correct proposed language is "All custom properties inherit from the originating element" combined with re-allowing custom properties on the highlight itself.

SGTM

> zig-zag

Too complicated IMO.

---

The resolution in #9909, ["On highlight pseudos, the var() function takes from the originating element."](https://github.com/w3c/csswg-drafts/issues/9909#issuecomment-1982141005), makes it sound like this is a special behavior for the var() lookup rather than actually affecting the computed values of the highlight style (the difference is visible in e.g. CSSOM). The IRC log does not show a discussion leading up to this conclusion; it kind of looks like we actually intended to affect the computed value?

In order to not make this more "special" than it needs to be, I do think we should affect the computed value rather than special-casing var() lookup. Especially if custom properties will actually be allowed on highlight pseudos again. Either way, it would be good to clarify this point.
